### DISCUSSION TITLE
Considering adding an option to opt-out from count for the list/search actions

### DISCUSSION THREAD

**Discussion by ganigeorgiev** [pocketbase/pocketbase#2965] [Announcements]
**TL;DR** With the next release there will be an optional query parameter to opt-out from the `totalItems` and `totalPages` pagination fields to speed up execution when the total number of collection records are not important. If this parameter is not explicitly set by default we will perform the count as it is currently. It will be explicitly disabled for the `getFullList()` and `getFirstListItem()` calls with the next PocketBase and JS SDK/Dart SDKs versions.

**The query parameter name (`withCount=1/0`, `disableCount=1/0`, etc.) is not yet decided, so feel free to comment if you have a suggestion for its naming.**

---

For the last couple days I've been experimenting, _unfortunatelly without much success_, with different approaches to try to optimize the `COUNT` query that is executed together with list/search actions (`getList()`, `getFullList()`, etc.) to generate the pagination count fields (`totalItems` and `totalPages`). This currently represents the main performance bottleneck when using the builtin PocketBase web APIs.

Most SQLite read queries are really fast. Performance seems to degrade when there are aggregations (`COUNT`, `SUM`, `GROUP BY`, etc.) used over large datasets.

This is not specific only for SQLite and the main reason for this is probably that aggregations like `COUNT` usually require performing a `SCAN` over the query table(s) and this is inherently slow for complex queries especially when multiple `JOIN`s with large tables are involved (_in PocketBase `JOIN` is peformed when there is a relation field lookup, eg. `someRelField.title = '123'`_).

The generated PocketBase queries when API rules and filters are used are almost identical to queries typed by hand and there is no much room for improvement in that regard (_there are some microoptimizations that we can apply when for example `someRelField.id = ...` is used that could be converted to `someRelField = ...` to avoid the extra `JOIN` but that will be optimized behind the scenes in the future_).

**Since in general there is nothing much we can do to speed up the `COUNT` execution (_or at least I couldn't find_), I'm planning with the next release to add an optional `?withCount=1/0` query parameter to the list/search action to opt-out and skip the `COUNT` entirely.**

This will be extremely useful when `getFullList()` and `getFirstListItem()` are used since in those cases the end-user usually don't care about the `totalItems` and `totalPages` response fields (_with `getFullList()` there is a chance we'll have to perform an extra request to determine when to stop making calls but overall the performance without the count is still better even with the extra get request_).

**When `withCount` is explicitly disabled the `totalItems` and `totalPages` will be either ignored from the result or set with negative `-1` value. Additionally, if count is enabled it will be executed concurrently with the regular `SELECT` one, meaning that we no longer will normalize the pagination offset and if the user request some `page` number like 9999999 we'll return empty results instead of the last available page as it is currently.**

**I haven't decided yet on the exact parameter naming and behavior of the total fields so feel free to comment with your suggestions.**

_Just for completeness, below is an example for the performance difference with and without a count for the 50k posts collection from the [benchmarks dataset](https://github.com/pocketbase/benchmarks):_

```
# with count (default)
posts50k - nested multiple relations lookup (no indexes, at-least-one) [reqs:100, conc:10, rule:`"author.permissions.active ?= true"`, query:`?perPage=20&withCount=1`]
┌─ Best:      3.848422472s
├─ Worst:     4.664202623s
├─ Completed: 42.29279228s
└─ Errors:    0

# no count
posts50k - nested multiple relations lookup (no indexes, at-least-one) [reqs:100, conc:10, rule:`"author.permissions.active ?= true"`, query:`?perPage=20&withCount=0`]
┌─ Best:      9.680025ms
├─ Worst:     110.778864ms
├─ Completed: 389.788187ms
└─ Errors:    0
```

**Comment by ganigeorgiev** [comment]
I decided on `skipTotal=1/0` for the parameter name. If the parameter is set then `totalItems` and `totalPages` will be returned as `-1` in the json response to indicate that the total counter was skipped.

The change is in the `develop` branch and will be available with the next release (the new defaults for `getFirstListItem()` and `getFullList()` in the JS and Dart SDKs were also updated and published as `rc1` releases if you want to test them).
The benchmark results will be updated sometime later next week.

I'm planning to release PocketBase v0.17.0 by the end of next week or the week following it.

**Comment by akashdeep000** [comment]
Another thing would be great if we can use  **Cursor-based** pagination instead of offset. 

- Reasons:

   - **Cursor-based** pagination solve **performance** issues that offset pagination has.
     
   - When using offset based pagination, it doesn't play well with real-time data or data that changes often. Offset says how many rows we want to skip but doesn't account for row deletion or new rows being created. Such an offset can result in showing duplicate data or some data missing. Where as Cursor-based pagination does not rely on the relative ordering of the Instead, it relies on an index created and managed by the database.

To implement cursor pagination, we will need the following information:

`limit` - Same as before, amount of rows we want to show on one page
`cursor` - ID of a reference element in the list. This can be the first item if you're querying the previous page and the last item if querying the next page.
`cursorDirection` - If user clicked Next or Previous (after or before)
When requesting the first page, we don't need to provide anything, just the limit 10, saying how many rows we want to get. As a result, we get our ten rows.

To get the next page, we use the ID of the last row as the `cursor` and set `cursorDirection` to after.

Similarly, if we want to go to the previous page, we use the ID of the first row as `cursor` and set `cursorDirection` to before.

To compare, in the SQL world, we could write our query as 
```sql
SELECT * FROM posts WHERE id > 10 LIMIT 10 ORDER BY id DESC
```

**Reply by ganigeorgiev** [reply]
Cursor based pagination should be already supported. You can find an example in https://github.com/pocketbase/pocketbase/issues/2680#issuecomment-1585629535.
